File: code-breaker/python/src/code_breaker/code_breaker.py
from dataclasses import dataclass
from enum import IntEnum
from typing import List, Tuple


class Peg(IntEnum):
    """The six playable values in a Mastermind-style code.

    Modeled as an ``IntEnum`` so invalid pegs cannot be constructed while
    equality, hashing, and list membership all behave like the underlying
    integer — which is what the scoring algorithm wants.
    """

    ONE = 1
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5
    SIX = 6


CODE_LENGTH = 4


PegTuple = Tuple[Peg, Peg, Peg, Peg]


@dataclass(frozen=True)
class Secret:
    pegs: PegTuple

    def __post_init__(self) -> None:
        if len(self.pegs) != CODE_LENGTH:
            raise ValueError(f"Secret must have exactly {CODE_LENGTH} pegs.")

    def score_against(self, guess: "Guess") -> "Feedback":
        return _compute_feedback(self.pegs, guess.pegs)


@dataclass(frozen=True)
class Guess:
    pegs: PegTuple

    def __post_init__(self) -> None:
        if len(self.pegs) != CODE_LENGTH:
            raise ValueError(f"Guess must have exactly {CODE_LENGTH} pegs.")


@dataclass(frozen=True)
class Feedback:
    exact_matches: int
    color_matches: int

    @property
    def is_won(self) -> bool:
        return self.exact_matches == CODE_LENGTH

    def render(self) -> str:
        return "+" * self.exact_matches + "-" * self.color_matches

    def __str__(self) -> str:
        return self.render()


def _compute_feedback(secret: PegTuple, guess: PegTuple) -> Feedback:
    """Two-pass scoring.

    Exact matches consume positions first; remaining guess pegs then pair
    with remaining secret pegs by value, each secret peg consumed at most
    once so duplicates never double-count.
    """
    exact_matches = 0
    unmatched_secret: List[Peg] = []
    unmatched_guess: List[Peg] = []

    for i in range(CODE_LENGTH):
        if secret[i] == guess[i]:
            exact_matches += 1
        else:
            unmatched_secret.append(secret[i])
            unmatched_guess.append(guess[i])

    color_matches = 0
    for peg in unmatched_guess:
        if peg in unmatched_secret:
            unmatched_secret.remove(peg)
            color_matches += 1

    return Feedback(exact_matches=exact_matches, color_matches=color_matches)

File: code-breaker/typescript/src/codeBreaker.ts
// The six playable values in a Mastermind-style code. A const-object-as-enum
// keeps the runtime values numeric (so comparisons are trivial) while giving
// TypeScript the literal-type safety that a raw `number` would lose.
export const Peg = {
  One: 1,
  Two: 2,
  Three: 3,
  Four: 4,
  Five: 5,
  Six: 6,
} as const;
export type Peg = (typeof Peg)[keyof typeof Peg];

export const CODE_LENGTH = 4;

export type CodePegs = readonly [Peg, Peg, Peg, Peg];

export interface Secret {
  readonly pegs: CodePegs;
  scoreAgainst(guess: Guess): Feedback;
}

export interface Guess {
  readonly pegs: CodePegs;
}

export interface Feedback {
  readonly exactMatches: number;
  readonly colorMatches: number;
  readonly isWon: boolean;
  render(): string;
}

export function createSecret(pegs: CodePegs): Secret {
  return {
    pegs,
    scoreAgainst(guess: Guess): Feedback {
      return computeFeedback(pegs, guess.pegs);
    },
  };
}

export function createGuess(pegs: CodePegs): Guess {
  return { pegs };
}

export function createFeedback(exactMatches: number, colorMatches: number): Feedback {
  return {
    exactMatches,
    colorMatches,
    isWon: exactMatches === CODE_LENGTH,
    render: () => '+'.repeat(exactMatches) + '-'.repeat(colorMatches),
  };
}

// Two-pass scoring: exact matches consume positions first, then the
// remaining secret pegs are matched against the remaining guess pegs
// by value, respecting multiplicity.
function computeFeedback(secret: CodePegs, guess: CodePegs): Feedback {
  let exactMatches = 0;
  const unmatchedSecret: Peg[] = [];
  const unmatchedGuess: Peg[] = [];

  for (let i = 0; i < CODE_LENGTH; i++) {
    if (secret[i] === guess[i]) {
      exactMatches++;
    } else {
      unmatchedSecret.push(secret[i] as Peg);
      unmatchedGuess.push(guess[i] as Peg);
    }
  }

  let colorMatches = 0;
  for (const peg of unmatchedGuess) {
    const index = unmatchedSecret.indexOf(peg);
    if (index >= 0) {
      unmatchedSecret.splice(index, 1);
      colorMatches++;
    }
  }

  return createFeedback(exactMatches, colorMatches);
}
